File: python/deepstrike/runtime/replay_provider.py
"""ReplayProvider — an LLMProvider that emits previously-recorded assistant messages
instead of calling a real LLM API.

Python port of node/src/runtime/replay-provider.ts. See that file for the full design
rationale. Distinct from `provider_replay` / `seed_provider_replay` which is the
session-repair reasoning-content cache (does NOT skip LLM calls).

Cost-accounting under replay:
- inputTokens estimated from the rendered context this call carries (NOT recorded).
- outputTokens taken from msg.token_count when present; else chars / 4.
- cacheReadInputTokens / cacheCreationInputTokens emitted as 0 — replay has no real
  cache state.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, AsyncIterator, Callable, Sequence

from deepstrike._kernel import Message, ToolSchema  # type: ignore
from deepstrike.providers.base import (
    LLMProvider,  # type: ignore[attr-defined]
    ProviderDescriptor,
    ProviderRunState,
    RenderedContext,
    StreamEvent,
)


@dataclass
class ReplayProviderOpts:
    """Options for ReplayProvider construction."""

    #: Maps a rendered-context text payload to a token count. Defaults to chars / 4.
    tokenizer: Callable[[str], int] | None = None
    #: Provider descriptor advertised via descriptor(). Defaults to a generic replay descriptor.
    descriptor: ProviderDescriptor | None = None
    #: When True, stream() and complete() wrap to the start once the fixture is exhausted.
    wrap: bool = False


_DEFAULT_DESCRIPTOR: ProviderDescriptor = {
    "provider": "replay",
    "protocol": "openai-chat",
    "model": "replay",
    "reasoning": {"supported": False, "preserveAcrossToolTurns": False},
    "toolCalls": {"supported": True, "requiresStrictPairing": False},
}


def _default_tokenizer(text: str) -> int:
    return (len(text) + 3) // 4


def _render_context_to_text(context: RenderedContext, tools: Sequence[ToolSchema]) -> str:
    parts: list[str] = []
    sys_text = getattr(context, "systemText", None) or context.get("systemText") if isinstance(context, dict) else getattr(context, "systemText", None)
    # Be tolerant of either dict-like RenderedContext or dataclass — use getattr / dict access.
    def _get(key: str) -> Any:
        if isinstance(context, dict):
            return context.get(key)
        return getattr(context, key, None)

    for key in ("systemText", "systemStable", "systemKnowledge"):
        v = _get(key)
        if v:
            parts.append(str(v))
    state_turn = _get("stateTurn")
    if state_turn:
        content = state_turn.get("content") if isinstance(state_turn, dict) else getattr(state_turn, "content", None)
        if content:
            parts.append(str(content))
    turns = _get("turns") or []
    for turn in turns:
        content = turn.get("content") if isinstance(turn, dict) else getattr(turn, "content", None)
        if content:
            parts.append(str(content))
        content_parts = turn.get("contentParts") if isinstance(turn, dict) else getattr(turn, "contentParts", None)
        for p in content_parts or []:
            for f in ("output", "text"):
                v = p.get(f) if isinstance(p, dict) else getattr(p, f, None)
                if isinstance(v, str):
                    parts.append(v)
                    break
        tool_calls = turn.get("toolCalls") if isinstance(turn, dict) else getattr(turn, "toolCalls", None)
        for tc in tool_calls or []:
            name = tc.get("name") if isinstance(tc, dict) else getattr(tc, "name", "")
            args = tc.get("arguments") if isinstance(tc, dict) else getattr(tc, "arguments", "")
            parts.append(f"{name} {args}")
    for t in tools or []:
        name = t.get("name") if isinstance(t, dict) else getattr(t, "name", "")
        desc = t.get("description") if isinstance(t, dict) else getattr(t, "description", "")
        params = t.get("parameters") if isinstance(t, dict) else getattr(t, "parameters", "")
        parts.append(f"{name} {desc} {params}")
    return "\n".join(parts)


class ReplayProvider(LLMProvider):
    """LLMProvider that dequeues recorded assistant messages instead of calling an API."""

    def __init__(self, messages: Sequence[Message], opts: ReplayProviderOpts | None = None) -> None:
        self._messages: tuple[Message, ...] = tuple(messages)
        opts = opts or ReplayProviderOpts()
        self._tokenizer = opts.tokenizer or _default_tokenizer
        self._descriptor = opts.descriptor or _DEFAULT_DESCRIPTOR
        self._wrap = bool(opts.wrap)
        self._cursor = 0

    def descriptor(self) -> ProviderDescriptor:
        return self._descriptor

    def consumed(self) -> int:
        return self._cursor

    def remaining(self) -> int:
        return max(0, len(self._messages) - self._cursor)

    def reset(self) -> None:
        self._cursor = 0

    def _pull(self) -> Message:
        if self._cursor >= len(self._messages):
            if self._wrap and self._messages:
                self._cursor = 0
            else:
                raise RuntimeError(
                    f"ReplayProvider: fixture exhausted (consumed={self._cursor}, total={len(self._messages)})"
                )
        msg = self._messages[self._cursor]
        self._cursor += 1
        return msg

    async def complete(
        self,
        context: RenderedContext,
        tools: list[ToolSchema],
        extensions: dict | None = None,
    ) -> Message:
        msg = self._pull()
        out: dict[str, Any] = {"role": "assistant", "content": getattr(msg, "content", "") or ""}
        tool_calls = getattr(msg, "toolCalls", None) or getattr(msg, "tool_calls", None)
        if tool_calls:
            out["toolCalls"] = tool_calls
        token_count = getattr(msg, "tokenCount", None) or getattr(msg, "token_count", None)
        if token_count is not None:
            out["tokenCount"] = token_count
        return out  # type: ignore[return-value]

    async def stream(  # type: ignore[override]
        self,
        context: RenderedContext,
        tools: list[ToolSchema],
        extensions: dict | None = None,
        state: ProviderRunState | None = None,
    ) -> AsyncIterator[StreamEvent]:
        msg = self._pull()
        content = getattr(msg, "content", None) or (msg.get("content") if isinstance(msg, dict) else "") or ""
        token_count = getattr(msg, "tokenCount", None) or (msg.get("tokenCount") if isinstance(msg, dict) else None)
        input_tokens = self._tokenizer(_render_context_to_text(context, tools))
        output_tokens = token_count if token_count is not None else self._tokenizer(content)

        yield {  # type: ignore[misc]
            "type": "usage",
            "totalTokens": input_tokens + output_tokens,
            "inputTokens": input_tokens,
            "outputTokens": output_tokens,
            "cacheReadInputTokens": 0,
            "cacheCreationInputTokens": 0,
        }
        if content:
            yield {"type": "text_delta", "delta": content}  # type: ignore[misc]

        tool_calls = getattr(msg, "toolCalls", None) or (msg.get("toolCalls") if isinstance(msg, dict) else None) or []
        import json

        for tc in tool_calls:
            name = tc.get("name") if isinstance(tc, dict) else getattr(tc, "name", "")
            tc_id = tc.get("id") if isinstance(tc, dict) else getattr(tc, "id", "")
            raw_args = tc.get("arguments") if isinstance(tc, dict) else getattr(tc, "arguments", "{}")
            try:
                args = json.loads(raw_args) if isinstance(raw_args, str) else (raw_args or {})
            except Exception:
                args = {}
            yield {"type": "tool_call", "id": tc_id, "name": name, "arguments": args}  # type: ignore[misc]

File: wasm/src/runtime/replay-provider.ts
/**
 * ReplayProvider — an LLMProvider that emits previously-recorded assistant messages
 * instead of calling a real LLM API.
 *
 * Purpose: deterministic re-runs for benchmarking, CI, and golden regression. Useful when you want
 * to hold the model's behavior constant and measure something else (prompt-size cost Δ across
 * RuntimeOptions variants, codegen of follow-on kernel work, etc.).
 *
 * Distinct from `provider-replay.ts`: that file's `seedProviderReplay` / `peekProviderReplay` is a
 * session-repair *reasoning-content cache* — it preserves `reasoning_content` / `native_blocks` so
 * the model sees its own thinking when context is re-rendered. It does NOT skip LLM calls.
 * `ReplayProvider` is the orthogonal, request-skipping mechanism: it returns recorded responses
 * directly, never hitting an API.
 *
 * Cost-accounting under replay:
 *   - `inputTokens` is ESTIMATED from the rendered context this call carries (NOT a recorded value
 *     from the original run). That's the point of replay-for-benchmarking: prompt may differ across
 *     variants, response is pinned, so a cost Δ purely reflects the prompt change.
 *   - `outputTokens` is taken from `message.tokenCount` when present; otherwise estimated from
 *     `message.content.length / 4`.
 *   - `cacheReadInputTokens` / `cacheCreationInputTokens` are emitted as 0 — replay has no real
 *     cache state. Mechanisms whose Δ depends on cache behavior must validate with a live A/B too.
 *
 * Tokenizer: by default a `chars/4` estimator (±20% for English; worse for code/JSON). For tighter
 * numbers plug `opts.tokenizer = tiktokenEncoder` or similar.
 */

import type {
  LLMProvider,
  Message,
  ProviderDescriptor,
  ProviderRunState,
  RenderedContext,
  StreamEvent,
  TextDelta,
  ToolCallEvent,
  ToolSchema,
  UsageEvent,
} from "../types.js"

export interface ReplayProviderOpts {
  /**
   * Maps a rendered-context text payload to a token count. Defaults to `chars / 4`.
   * Pass a real encoder (tiktoken etc.) for accurate cost accounting under replay.
   */
  tokenizer?: (text: string) => number
  /**
   * Provider descriptor advertised via `descriptor()`. Defaults to a generic
   * `{ provider: "replay", protocol: "replay", ... }` shape. Override when a downstream consumer
   * needs to detect the original provider (e.g., for protocol-specific decoding paths).
   */
  descriptor?: ProviderDescriptor
  /**
   * When true, `stream()` and `complete()` wrap to the start once the fixture is exhausted,
   * instead of throwing. Useful for loop tests that need to keep going past the recorded length.
   * Defaults to false.
   */
  wrap?: boolean
}

const DEFAULT_DESCRIPTOR: ProviderDescriptor = {
  provider: "replay",
  protocol: "openai-chat",
  model: "replay",
  reasoning: { supported: false, preserveAcrossToolTurns: false },
  toolCalls: { supported: true, requiresStrictPairing: false },
}

export class ReplayProvider implements LLMProvider {
  private cursor = 0
  private readonly messages: ReadonlyArray<Message>
  private readonly tokenizer: (text: string) => number
  private readonly _descriptor: ProviderDescriptor
  private readonly wrap: boolean

  /**
   * @param messages Ordered list of assistant messages to replay (one per LLM call).
   * @param opts Optional tokenizer / descriptor / wrap-around behavior.
   */
  constructor(messages: ReadonlyArray<Message>, opts: ReplayProviderOpts = {}) {
    this.messages = messages
    this.tokenizer = opts.tokenizer ?? defaultTokenizer
    this._descriptor = opts.descriptor ?? DEFAULT_DESCRIPTOR
    this.wrap = !!opts.wrap
  }

  descriptor(): ProviderDescriptor {
    return this._descriptor
  }

  /** Number of messages consumed so far. */
  consumed(): number {
    return this.cursor
  }

  /** Number of messages remaining in the fixture (returns 0 in wrap mode once cursor passes end). */
  remaining(): number {
    return Math.max(0, this.messages.length - this.cursor)
  }

  /** Reset the cursor — useful for re-running the same fixture in a fresh session. */
  reset(): void {
    this.cursor = 0
  }

  async complete(_context: RenderedContext, _tools: ToolSchema[]): Promise<Message> {
    const msg = this.pull()
    return {
      role: "assistant",
      content: msg.content,
      ...(msg.toolCalls ? { toolCalls: msg.toolCalls } : {}),
      ...(msg.tokenCount !== undefined ? { tokenCount: msg.tokenCount } : {}),
    }
  }

  async *stream(
    context: RenderedContext,
    tools: ToolSchema[],
    _extensions?: Record<string, unknown>,
    _state?: ProviderRunState,
    _signal?: AbortSignal,
  ): AsyncIterable<StreamEvent> {
    const msg = this.pull()
    const inputTokens = this.estimateInputTokens(context, tools)
    const outputTokens =
      msg.tokenCount !== undefined ? msg.tokenCount : this.tokenizer(msg.content || "")

    const usage: UsageEvent = {
      type: "usage",
      totalTokens: inputTokens + outputTokens,
      inputTokens,
      outputTokens,
      cacheReadInputTokens: 0,
      cacheCreationInputTokens: 0,
    }
    yield usage

    if (msg.content) {
      const delta: TextDelta = { type: "text_delta", delta: msg.content }
      yield delta
    }

    for (const tc of msg.toolCalls ?? []) {
      let args: Record<string, unknown> = {}
      try {
        args = JSON.parse(tc.arguments || "{}")
      } catch {
        // Malformed recorded arguments — pass an empty object. The runner's downstream tool
        // execution will surface the error if the tool needs them.
      }
      const call: ToolCallEvent = { type: "tool_call", id: tc.id, name: tc.name, arguments: args }
      yield call
    }
  }

  private pull(): Message {
    if (this.cursor >= this.messages.length) {
      if (this.wrap && this.messages.length > 0) {
        this.cursor = 0
      } else {
        throw new Error(
          `ReplayProvider: fixture exhausted (consumed=${this.cursor}, total=${this.messages.length})`,
        )
      }
    }
    return this.messages[this.cursor++]
  }

  private estimateInputTokens(context: RenderedContext, tools: ToolSchema[]): number {
    const text = renderContextToText(context, tools)
    return this.tokenizer(text)
  }
}

// ── helpers ───────────────────────────────────────────────────────────────────

function defaultTokenizer(text: string): number {
  return Math.ceil(text.length / 4)
}

function renderContextToText(context: RenderedContext, tools: ToolSchema[]): string {
  const parts: string[] = []
  if (context.systemText) parts.push(context.systemText)
  if (context.systemStable) parts.push(context.systemStable)
  if (context.systemKnowledge) parts.push(context.systemKnowledge)
  if (context.stateTurn?.content) parts.push(context.stateTurn.content)
  for (const turn of context.turns ?? []) {
    if (turn.content) parts.push(turn.content)
    for (const part of turn.contentParts ?? []) {
      const p = part as unknown as Record<string, unknown>
      if (typeof p.output === "string") parts.push(p.output)
      else if (typeof p.text === "string") parts.push(p.text)
    }
    for (const tc of turn.toolCalls ?? []) {
      parts.push(`${tc.name} ${tc.arguments}`)
    }
  }
  for (const tool of tools) {
    parts.push(`${tool.name} ${tool.description} ${tool.parameters}`)
  }
  return parts.join("\n")
}
